engine/execution/state/cargo/queue/queue.go

```diff
@@ -0,0 +1,97 @@
+package queue
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type headerInContext struct {
+	Header *flow.Header
+	ID     flow.Identifier
+}
+
+// FinalizedBlockQueue is a concurrency-safe queue of finalized block headers
+// Every time a new header is added to the queue it verifies the complience of the header to the
+// most recenlty added header.
+type FinalizedBlockQueue struct {
+	headers            []headerInContext
+	lock               sync.RWMutex
+	lastDequeuedHeader *headerInContext
+	lastQueuedHeader   *headerInContext
+}
+
+// NewFinalizedBlockQueue constructs a new FinalizedBlockQueue
+// "genesis“ is not inserted in the queue and is only used to to validate the very first incoming header
+func NewFinalizedBlockQueue(
+	genesis *flow.Header,
+) *FinalizedBlockQueue {
+	return &FinalizedBlockQueue{
+		headers:            make([]headerInContext, 0),
+		lastDequeuedHeader: &headerInContext{genesis, genesis.ID()},
+	}
+}
+
+// Enqueue checks the header compatibility and append a header to the queue
+// A header is compatible if its parentID matches the ID of the last inserted header (or genesis)
+// and its height is set right (last block's height plus one).
+// an error is returned if header not compatible
+func (ft *FinalizedBlockQueue) Enqueue(header *flow.Header) error {
+	ft.lock.Lock()
+	defer ft.lock.Unlock()
+
+	expectedHeight := ft.lastDequeuedHeader.Header.Height + 1
+	expectedParentID := ft.lastDequeuedHeader.ID
+	if len(ft.headers) > 0 {
+		expectedHeight = ft.lastQueuedHeader.Header.Height + 1
+		expectedParentID = ft.lastQueuedHeader.ID
+	}
+	if expectedHeight != header.Height || expectedParentID != header.ParentID {
+		return &NonCompliantHeaderError{
+			expectedHeight,
+			header.Height,
+			expectedParentID,
+			header.ParentID,
+		}
+	}
+
+	h := headerInContext{header, header.ID()}
+	ft.headers = append(ft.headers, h)
+	ft.lastQueuedHeader = &h
+	return nil
+}
+
+// Peak returns the oldest header in the queue without removing it
+// If the queue is empty it returns zero ID and nil header
+func (ft *FinalizedBlockQueue) Peak() (flow.Identifier, *flow.Header) {
```

Comment:
Thats intentional, the `Peek` is needed for synchronization of headers with state ([here](https://github.com/onflow/flow-go/pull/4249/files#diff-4138a9e8ec5d09ce7fef48ef48fec6e7f6e53160f326d4d0d2315f0a60c21834R93-R102)), the state might be behind and we don't want to consume the headers anymore.

Reply:
The exposed api is "thread safe", but not data race safe.

Consider the following scenario:

goroutine 1: peek (block 1), do some work, then context switch to goroutine 2
goroutine 2: peek (block 1), do some work, dequeue (block 1), then context switch back to goroutine 1
goroutine 1: dequeue (some other block!!!)

This is by definition a byzantine failure (Removing Peek and HasHeaders eliminate this data race).

If the queue only support multiple producer / single consumer, then please explicitly comment the safety model, and maybe make the data structure private so that it could not be accidentally used by other code bases.